File: Algo/Leetcode/133CloneGraph.py
"""
# Definition for a Node.
class Node(object):
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""


class Solution(object):

    def __init__(self):
        self.visit = {}  # old node to new node

    def cloneGraph(self, node):
        """
        :type node: Node
        :rtype: Node
        """
        return self.dfs(node)

    def dfs(self, node):
        if node is None:
            return node

        if node in self.visit:
            return self.visit[node]  # incase dfs repeatly

        clone_node = Node(node.val, [])
        self.visit[node] = clone_node
        for n in node.neighbors:
            clone_node.neighbors.append(self.dfs(n))

        return clone_node

File: Algo/Leetcode/133CloneGraph.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/

class Solution {
public:
    map<Node*, Node*> visit;
    Node* dfs(Node* node) {
        if (!node)
            return NULL;

        if (visit.find(node) != visit.end())
            return visit[node];

        Node* clone_node = new Node(node->val);
        visit[node] = clone_node;

        for (auto p: node->neighbors) {
            clone_node->neighbors.push_back(dfs(p));
        }

        return clone_node;
    }
    Node* cloneGraph(Node* node) {
        return dfs(node);

    }
};